### PR TITLE
make validatingadmissionpolicy.policyDecision public for reuse

### PR DESCRIPTION
**Pull request by JimBugwadia**
Signed-off-by: Jim Bugwadia <jim@nirmata.com>

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Make [validating.policyDecision](https://pkg.go.dev/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy#Validator) returned via the Validator interface public to allow reuse in external systems. 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/policy_decision.go [hunk 1]
```diff
@@ -24,33 +24,33 @@ import (
...
+	Message    string
+	Reason     metav1.StatusReason
+	Elapsed    time.Duration
 }
 
-type policyDecisionWithMetadata struct {
-	policyDecision
-	definition *v1alpha1.ValidatingAdmissionPolicy
-	binding    *v1alpha1.ValidatingAdmissionPolicyBinding
+type PolicyDecisionWithMetadata struct {
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/policy_decision.go:50]: `PolicyDecisionWithMetadata` is not part of external api and I think should remain internal

**Review by alexzielenski — CHANGES_REQUESTED** [review]
I support this in theory I think. But I'm curious what you are doing with these types? Are you making a custom policy validator?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/policy_decision.go (hunk 1 above)
Comment (JimBugwadia) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/policy_decision.go]: Ok, was not sure on `PolicyDecisionWithMetadata`. Have switched back to private.

**Comment by JimBugwadia** [comment]
> I support this in theory I think. But I'm curious what you are doing with these types? Are you making a custom policy validator?

Thanks! I am exploring ideas for how to integrate with Kyverno and was trying to apply a ValidatingAdmissionPolicy to a resource in a standalone program.

**Comment by JimBugwadia** [comment]
Hi @alexzielenski - any additional comments or questions?

**Comment by alexzielenski** [comment]
/approve

/assign @cici37 

 for lgtm

**Comment by JimBugwadia** [comment]
Hi @cici37 - thanks for accepting to review! Please let me know if you any feedback, or recommend any changes.

**Comment by cici37** [comment]
Thanks for the PR! Would you mind providing further info on why this is needed for external use or if you could use the existing returned info from validatingadmissionpolicy? (The reason I asked is that we generally tend to not to export stuff for external usage since we don't guarantee the consistency for internal interface/struct/fields etc.)

**Comment by JimBugwadia** [comment]
Thanks for reviewing!

I would like to use the Validator interface to apply policies to a resource. 

Currently, the Validate methods and all it's inputs are public, but the returned results ([]policyDecision) was not.

Should the result also be an interface? I thought about that, but it would be a more significant change. Happy to pursue that if you think it's a better approach.

**Comment by cici37** [comment]
/lgtm
/hold
in case @maxsmythe wanna take a look

Thank you!

**Comment by maxsmythe** [comment]
@cici37 

I'm okay with this or #115412 

However, if this get's merged, we'll still need the second part of #115412, which separates version conversion from validation, since version conversion is only reliably available in the API server.

**Comment by cici37** [comment]
@JimBugwadia  Extra info: we currently don't return all PolicyDecisions back to client(we only return the first rejection if I remember correctly). So if you plan to use PolicyDecision, it might cause inconsistency if that is a concern. This behavior is not guaranteed to stay the same since it is an alpha feature but at least currently we settled as return the first rejection. The metrics counts will be increased by multiple violations though.

**Comment by JimBugwadia** [comment]
Looks like https://github.com/kubernetes/kubernetes/pull/115412 includes changes I made. Closing this PR.

**Pull request closed without merging**